docs/lang/articles/misc/ndarray_android.md

```diff
@@ -0,0 +1,449 @@
+
+Taking a simulation of celestial bodies' orbits as an example, this tutorial walks you through the process of running a Taichi program using Ndarray on Android.
+
+> - Taichi kernels and compute shaders are *not* one-to-one mapping. Each Taichi kernel can generate multiple compute shaders, the number *usually* in line with the number of the for loops in the kernel.
+> - [Taichi's AOT (Ahead Of Time) module](https://github.com/taichi-dev/taichi/issues/3642) is currently a proof of concept under development and subject to change in the future.
+
+## A definition of Ndarray
+
+Taichi provides a data container called Ndarray.  An Ndarray is a multidimensional container of elements of the same type and size; an element in an Ndarray is virtually a scalar or a tensor.
+
+### Ndarray shape
+
+Ndarray shape defines the Ndarray's layout; element shape defines the element's layout. For example:
+
+- An Ndarray with an Ndarray shape of [2, 1024] and an element shape of [1] is an array of 2 x 1,024 = 2,048 scalars.
+- An Ndarray with an Ndarray shape of [128, 128] and an element shape of [2, 4] is an array of  128 x 128 = 16,384 2 by 4 matrices.
+
+### Ndarray dimension
+
+The dimension here refers to the number of dimensions of an Ndarray. For example:
+
+- The dimension of an Ndarray with an Ndarray shape of [1, 2, 3] is three.
+- The dimension of an Ndarray with an Ndarray shape of [500] is one.
+
+### Benefit of Ndarray
+
+Each Ndarray has a fixed dimension but gives you the flexibility of changing its shape in accordance with its dimension.
+
+Unlike a field's shape, which requires you to rewrite and recompile your Taichi program once it is changed, an Ndarray's shape can be *dynamically* changed without the need to recompile.
+
+Taking the simulation of celestial bodies' orbits as an example, suppose you wish to double the number of your celestial bodies to 2,000:
+
+- With Taichi field, you have to compile twice;
+- With Ndarray, all you need to do is to update your runtime program.
+
+## Run a Taichi program using Ndarray on Android
+
+The following section walks you through the process of running a Taichi program using Ndarray on Android.
+
+1. [Generate necessary compute shaders](#generate-necessary-compute-shaders)
+2. [Parse the generated JSON file](#parse-the-generated-json-file)
+3. [Prepare SSBO and shape information](#prepare-ssbo-and-shape-information)
+4. [Prepare rendering shaders](#prepare-rendering-shaders)
+5. [Execute all shaders](#execute-all-shaders)
+
+:::note
+
+From Step 2, you are required to come up with your own runtime program. We provide an [example Java runtime program for Android](https://github.com/taichi-dev/taichi-aot-demo/blob/master/nbody_ndarray/java_runtime/NbodyNdarray.java) for your reference, but you may need to adapt these codes for your platform and programming language.
```

Comment:
The java demo URL is inaccessible -- the taichi-aot-demo repository is private.

Reply:
We somehow overlooked this. Great to have a fresh pair of eyes on this! @Leonz5288 has just made the repo public. 

Thanks again!